### PR TITLE
Add telemetry as a level

### PR DESCRIPTION
**Pull request by AbeetoChanito**
#### Summary
Add telemetry as a level.

#### Motivation
Telemetry data needs to have its own level because telemetry data can't really have a severity. It's literally just data.

## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/6983240271)
- via manual download: [LemLib@0.5.0-rc1-c904646.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1071856355.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc1-c904646.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1071856355.zip;
pros c fetch LemLib@0.5.0-rc1-c904646.zip;
pros c apply LemLib@0.5.0-rc1-c904646;
rm LemLib@0.5.0-rc1-c904646.zip;
```

## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/7660299214)
- via manual download: [LemLib@0.5.0-rc2+dfb6bc.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1195995920.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc2+dfb6bc.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1195995920.zip;
pros c fetch LemLib@0.5.0-rc2+dfb6bc.zip;
pros c apply LemLib@0.5.0-rc2+dfb6bc;
rm LemLib@0.5.0-rc2+dfb6bc.zip;
```

### CODE REVIEW

File: include/lemlib/logger/message.hpp [hunk 1]
```diff
@@ -7,7 +7,7 @@ namespace lemlib {
  * @brief Level of the message
  *
  */
-enum class Level { INFO, DEBUG, WARN, ERROR, FATAL };
+enum class Level { INFO, DEBUG, WARN, ERROR, FATAL, TELEMETRY };
```
Comment (liam-teale) [inline include/lemlib/logger/message.hpp]: Shouldn't telemetry be before info?

File: include/lemlib/logger/buffer.hpp [hunk 2]
```diff
@@ -30,12 +32,11 @@ class Buffer {
         Buffer(const Buffer&) = delete;
         Buffer& operator=(const Buffer&) = delete;
 
-        /**
-         * @brief Push to the buffer
-         *
-         * @param bufferData
-         */
-        void pushToBuffer(const std::string& bufferData);
+        template <typename... T> void pushToBuffer(fmt::format_string<T...> format, T&&... args) {
```
Comment (liam-teale) [inline include/lemlib/logger/buffer.hpp:36]: where's the doxygen comment block?

File: include/lemlib/logger/telemetrySink.hpp [hunk 3]
```diff
@@ -23,6 +23,18 @@ class TelemetrySink : public BaseSink {
          * @brief Construct a new Telemetry Sink object
          */
         TelemetrySink();
+
+        /**
+         * @brief Log a message at the telemetry level
+         * @tparam T
+         * @param format
+         * @param args
+         */
```
Comment (liam-teale) [inline include/lemlib/logger/telemetrySink.hpp]: Document what the parameters do here

File: include/lemlib/logger/telemetrySink.hpp [hunk 4]
```diff
@@ -23,6 +23,18 @@ class TelemetrySink : public BaseSink {
...
+        template <typename... T> void warn(fmt::format_string<T...> format, T&&... args) {
+            if (isCombinedSink()) return;
+
```
Comment (liam-teale) [inline include/lemlib/logger/telemetrySink.hpp:35]: whats up with this breakline? Its a 2-line function

File: src/lemlib/logger/baseSink.cpp [hunk 5]
```diff
@@ -18,5 +18,7 @@ fmt::dynamic_format_arg_store<fmt::format_context> BaseSink::getExtraFormattingA
     return {};
 }
 
+bool BaseSink::isCombinedSink() { return !sinks.empty(); }
```
Comment (liam-teale) [inline src/lemlib/logger/baseSink.cpp:21]: missing doxygen comment block

File: src/lemlib/logger/buffer.cpp [hunk 6]
```diff
@@ -21,12 +21,6 @@ Buffer::~Buffer() {
     while (!buffersEmpty()) { pros::delay(10); }
 }
 
-void Buffer::pushToBuffer(const std::string& bufferData) {
-    mutex.take();
-    buffer.push_back(bufferData);
-    mutex.give();
-}
-
 void Buffer::setRate(uint32_t rate) { this->rate = rate; }
```
Comment (liam-teale) [inline src/lemlib/logger/buffer.cpp:18]: no doxygen comment block

File: src/lemlib/logger/buffer.cpp [hunk 7]
```diff
@@ -21,12 +21,6 @@ Buffer::~Buffer() {
...
 
 void Buffer::taskLoop() {
```
Comment (liam-teale) [inline src/lemlib/logger/buffer.cpp:20]: no doxygen comment block

File: include/lemlib/logger/stdout.hpp [hunk 8]
```diff
@@ -22,7 +22,7 @@ class BufferedStdout : public Buffer {
          *
          */
```
Comment (liam-teale) [inline include/lemlib/logger/stdout.hpp:26]: No parameters?

⠀⣞⢽⢪⢣⢣⢣⢫⡺⡵⣝⡮⣗⢷⢽⢽⢽⣮⡷⡽⣜⣜⢮⢺⣜⢷⢽⢝⡽⣝
⠸⡸⠜⠕⠕⠁⢁⢇⢏⢽⢺⣪⡳⡝⣎⣏⢯⢞⡿⣟⣷⣳⢯⡷⣽⢽⢯⣳⣫⠇
⠀⠀⢀⢀⢄⢬⢪⡪⡎⣆⡈⠚⠜⠕⠇⠗⠝⢕⢯⢫⣞⣯⣿⣻⡽⣏⢗⣗⠏⠀
⠀⠪⡪⡪⣪⢪⢺⢸⢢⢓⢆⢤⢀⠀⠀⠀⠀⠈⢊⢞⡾⣿⡯⣏⢮⠷⠁⠀⠀
⠀⠀⠀⠈⠊⠆⡃⠕⢕⢇⢇⢇⢇⢇⢏⢎⢎⢆⢄⠀⢑⣽⣿⢝⠲⠉⠀⠀⠀⠀
⠀⠀⠀⠀⠀⡿⠂⠠⠀⡇⢇⠕⢈⣀⠀⠁⠡⠣⡣⡫⣂⣿⠯⢪⠰⠂⠀⠀⠀⠀
⠀⠀⠀⠀⡦⡙⡂⢀⢤⢣⠣⡈⣾⡃⠠⠄⠀⡄⢱⣌⣶⢏⢊⠂⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢝⡲⣜⡮⡏⢎⢌⢂⠙⠢⠐⢀⢘⢵⣽⣿⡿⠁⠁⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠨⣺⡺⡕⡕⡱⡑⡆⡕⡅⡕⡜⡼⢽⡻⠏⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⣼⣳⣫⣾⣵⣗⡵⡱⡡⢣⢑⢕⢜⢕⡝⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⣴⣿⣾⣿⣿⣿⡿⡽⡑⢌⠪⡢⡣⣣⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⡟⡾⣿⢿⢿⢵⣽⣾⣼⣘⢸⢸⣞⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠁⠇⠡⠩⡫⢿⣝⡻⡮⣒⢽⠋⠀⠀⠀

**Review by liam-teale — CHANGES_REQUESTED** [review]
Sound idea. definitely needed. Clean it up, test, and i'll merge it in

File: include/lemlib/logger/telemetrySink.hpp (hunk 3 above)
Comment (AbeetoChanito) [reply include/lemlib/logger/telemetrySink.hpp]: I documented everything now, so this shouldn't be an issue for most of the logging codebase anymore.

**Review by liam-teale — APPROVED** [review]
LGTM. Will merge upon testing

**Comment by liam-teale** [comment]
@AbeetoChanito you are aware this does not compile right

**Comment by AbeetoChanito** [comment]
This has been open for far too long without any resolution. Also, I'm going to write LemLog anyway so this PR is not necessary for the release of 0.5.0. I'm going to close this PR.

**Pull request closed without merging**